src/main/java/org/javarosa/core/model/TriggerableDag.java

```diff
@@ -344,26 +343,31 @@ public void finalizeTriggerables(FormInstance mainInstance) throws IllegalStateE
     }
 
     private static Set<QuickTriggerable> buildDag(Set<QuickTriggerable> vertices, List<QuickTriggerable[]> edges) {
+        // The dag and the set of remaining vertices will be mutated
+        // inside the while loop's block
         Set<QuickTriggerable> dag = new LinkedHashSet<>();
-
         Set<QuickTriggerable> remainingVertices = new HashSet<>(vertices);
+
+        // The set of remaining edges will be replaced inside
+        // the while loop's block
         Set<QuickTriggerable[]> remainingEdges = new HashSet<>(edges);
+
         while (remainingVertices.size() > 0) {
+            // Compute the set of roots (nodes that don't show up
+            // as edge targets) with the remaining vertices
             Set<QuickTriggerable> roots = new HashSet<>(remainingVertices);
             for (QuickTriggerable[] edge : remainingEdges)
                 roots.remove(edge[1]);
 
             if (roots.size() == 0)
                 throwCyclesInDagException(vertices);
 
+            // "Move" the roots detected during this iteration
```

Comment:
With your new comments, this whole method feels pretty comfortable to understand to me, now. 👍